In [1]:
import pandas as pd
import numpy as np

In [2]:
df_scores = pd.read_csv('https://raw.githubusercontent.com/kinpkt/catlover-cs131-practice/refs/heads/main/practice_dataset/calculus_student_scores.csv')
df_scores

calculus_scores  student_count
0           30-40              1
1           41-50              1
2           51-60              9
3           61-70             19
4           71-80             22
5           81-90              5
6          91-100              3

**Part 1 : หาค่าเฉลี่ยเลขคณิตของคะแนน**

In [3]:
n = sum(df_scores.student_count)
boundary = df_scores.calculus_scores.str.split('-', expand=True).astype(int)
midpoint = (boundary[0] + boundary[1])/2

mean_scores = sum(df_scores.student_count*midpoint)/n
print(f'Mean = {mean_scores}')

Mean = 69.99166666666666


**Part 2 : หาค่ามัธยฐานของคะแนน**

In [4]:
n_half = n/2
df_scores['cum_students'] = df_scores.student_count.cumsum()
e_med = 3
Le_mean = 61-0.5
I = np.ceil(70.5-60.5)
fe_med = df_scores.student_count[e_med]
fb_med = df_scores.cum_students[e_med-1]

med_scores = Le_mean + (n_half-fb_med)/fe_med*I

print(f'Median = {med_scores}')

Median = 70.5


**Part 3 : หาค่าฐานนิยมของคะแนน**

In [5]:
e_mode = 4
Le_mode = 71-0.5
d_prev = np.abs(df_scores.student_count[e_mode] - df_scores.student_count[e_mode-1])
d_next = np.abs(df_scores.student_count[e_mode] - df_scores.student_count[e_mode+1])

mode_scores = Le_mode + d_prev/(d_prev+d_next)*I
print(f'Mode = {mode_scores}')

Mode = 72.0


**Part 4 : "หากในวิชานี้ตัด F จากคะแนนที่ต่ำกว่า Q<sub>2</sub> จะมีนิสิตที่ได้ F ประมาณกี่คน"**

In [6]:
Q2_index = n*2/4
e_Q2 = 3
Le_Q2 = 61-0.5

Q2 = Le_Q2 + (Q2_index-df_scores.cum_students[e_Q2-1])/df_scores.student_count[e_Q2]*I
print(f'Q2 = {Q2}') # ปัดขึ้นได้เป็น Q2 = 71

failed_students = 0

for i in range(len(boundary[1])):
    if boundary[1][i] < 71:
        failed_students += df_scores.student_count[i]

print(f'{failed_students} students got F on Calculus.')

Q2 = 70.5
30 students got F on Calculus.


**Part 5 : ถ้านิสิตที่ได้คะแนนมากกว่า P<sub>85</sub> หรือได้ 81 คะแนนขึ้นไปจะได้ A จะมีนิสิตที่ได้ A ประมาณกี่คน**

In [7]:
P85_index = n*85/100
e_P85 = 4
Le_P85 = 71-0.5

P85 = Le_P85 + (P85_index-df_scores.cum_students[e_P85-1])/df_scores.student_count[e_P85]*I
print(f'P85 = {P85}') # ปัดลงได้เป็น P85 = 80

a_students = 0

for i in range(len(boundary[1])):
    if boundary[1][i] > 80 or boundary[1][i] >= 81:
        a_students += df_scores.student_count[i]

print(f'{a_students} students got A on Calculus.')

P85 = 80.04545454545455
8 students got A on Calculus.


**Part 6 : การแจกแจงของคะแนนนิสิตทั้งหมดที่อยู่ในชุดข้อมูลนี้มีลักษณะอย่างไร**

In [8]:
df_scores

calculus_scores  student_count  cum_students
0           30-40              1             1
1           41-50              1             2
2           51-60              9            11
3           61-70             19            30
4           71-80             22            52
5           81-90              5            57
6          91-100              3            60

จากตาราง จะเห็นได้ว่าจำนวนนิสิตส่วนมาก ได้คะแนนในช่วงอันตรภาคชั้นที่ 4 และ 3 ตามลำดับ และจากการวิเคราะห์ข้อมูลเบื้องต้น เราจะสามารถสรุปได้ดังตารางว่า
|           |   Value  |
|:---------:|:--------:|
| $\bar{x}$ | $69.992$ |
|   $M_e$   |  $70.5$  |
|   $M_o$   |  $72.0$  |

จะเห็นได้ว่า ทั้งสามค่าที่ได้นั้นมีค่าใกล้เคียงกัน โดย $\bar{x}\approx M_e$ จึงสรุปได้ว่าการแจกแจงข้อมูลคะแนนของนิสิตมีการแจกแจงแบบเบ้ซ้ายและใกล้เคียงกับ Normal Distribution มาก